### ISSUE TITLE
Improve Router

### ISSUE THREAD

**Issue by vladimino** [MagedAhmad/PHP-microframework#21]
## Context

Currently, `Router` is responsible for too many things. It knows, how to:

1. Load configuration file
2. Parse request parameters
3. Instantiate controller and call specific action
4. Display error 404  page

This drastically violates the Single Responsibility Principle and requires some rethinking of its design.

Another problem is that configuration requires interaction with a `$router` object, e.g. `$router->get('', 'HomeController@home');`. This is not a good practice either, configuration should be abstract.

## Proposal

1. *Make config format abstract*

There must be no specific router information. Ideally, it should be YAML or some other text format, but it could be as simple as PHP array:

```php
$routes = [
  '' => 'HomeController@home'
];
```

or similar to an internal `$routes` property structure:

```php
$routes = [
    'GET' => [
        '' => 'HomeController@home',
    ],
    'POST' => [],
];
```

2. *Pass config values to `Router` from the outside*

Class `Router` must be instantiated with a configuration array passed to the constructor and shouldn't know where these values are coming from (either from the file or anything else).

3. *Do all request specific logic in the `Request` class*

At first, `Router` should have `Request` as a dependency, don't instantiate it directly within `get()` method. Having `new` keyword anywhere outside of creational classes it is actually a code smell.

Next, implement parameter parsing (all work that is done on [lines 19-29 of current `Router` class](https://github.com/MagedAhmad/PHP-microframework/blob/39a5d2bc8490984b9bc07edb1df701714144394f/src/Core/Router.php#L19-L29) in the `Request` class and expose it through a public method, let's say, `getSlug()`. Then you just call this method and check in your route configurations if there's one for the given slug (`''` in case of home page).

4. *Don't call `view()` function within `Router`*

`Router` should have one job: direct user to appropriate controller if a route is found or throw an error if not. You did well in case when `$action` is not found but not when there's no such route. Throwing an exception there and handling this outside of the`Router` would be a much cleaner solution. 

By the way, `\Exception` is too generic and it is always better to use as specific error as possible. In this case most suitable would be `\InvalidArgumentException` because controller and action names are coming from the URI as arguments.

## To-do

To summarize, this issue will be considered done when all these subtasks are solved:

- [x] Abstract configuration format
- [x] Pass configuration values from the outside of `Router`
- [x] Inject `Request` to `Router`
- [x] Implement method to get URI slug without query parameters in the `Request` class
- [x] Throw specific error message if route is not found in `Router`
- [x] Catch this exception and display 404 page to handle the situation with a good user experience
